[lldb] Make use of Scripted{Python,}Interface for ScriptedThreadPlan

medismailben opened the pull request:
This patch makes ScriptedThreadPlan conforming to the ScriptedInterface
& ScriptedPythonInterface facilities by introducing 2
ScriptedThreadPlanInterface & ScriptedThreadPlanPythonInterface classes.

This allows us to get rid of every ScriptedThreadPlan-specific SWIG
method and re-use the same affordances as other scripting offordances,
like Scripted{Process,Thread,Platform} & OperatingSystem.

To do so, this adds new transformer methods for `ThreadPlan`, `Stream` &
`Event`, to allow the bijection between C++ objects and their python
counterparts.

bulbazord commented on lldb/include/lldb/Interpreter/Interfaces/ScriptedThreadPlanInterface.h:
```diff
@@ -0,0 +1,40 @@
+//===-- ScriptedThreadInterface.h -------------------------------*- C++ -*-===//
```
The header is missing the world `Plan`

bulbazord commented on lldb/include/lldb/Interpreter/Interfaces/ScriptedThreadPlanInterface.h:
```diff
@@ -0,0 +1,40 @@
+namespace lldb_private {
+class ScriptedThreadPlanInterface : public ScriptedInterface {
+public:
+  virtual llvm::Expected<StructuredData::GenericSP>
+  CreatePluginObject(llvm::StringRef class_name,
+                     lldb::ThreadPlanSP thread_plan_sp,
+                     const StructuredDataImpl &args_sp) {
+    llvm_unreachable("unimplemented!");
+  }
+
+  virtual llvm::Expected<bool> ExplainsStop(Event *event) { return true; }
+
+  virtual llvm::Expected<bool> ShouldStop(Event *event) { return true; }
+
+  virtual llvm::Expected<bool> IsStale() { return true; };
+
+  virtual lldb::StateType GetRunState() { return lldb::eStateStepping; }
+
+  virtual llvm::Expected<bool> GetStopDescription(lldb_private::Stream *s) {
+    return true;
```
Why not make all of these pure virtual methods?

medismailben replied:
Same answer as in https://github.com/llvm/llvm-project/pull/68052#discussion_r1372058327.

bulbazord commented on lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedThreadPlanPythonInterface.cpp:
```diff
@@ -0,0 +1,92 @@
+
+lldb::StateType ScriptedThreadPlanPythonInterface::GetRunState() {
+  Status error;
+  StructuredData::ObjectSP obj = Dispatch("should_step", error);
```
You don't need to do anything but `should_step` is so close to `should_stop`... 😱

medismailben replied:
Yep, but that's stable API at this point, can't change it at this time 😭

bulbazord commented on lldb/source/Interpreter/ScriptInterpreter.cpp:
```diff
@@ -104,6 +104,19 @@ ScriptInterpreter::GetStatusFromSBError(const lldb::SBError &error) const {
+Stream *ScriptInterpreter::GetOpaqueTypeFromSBStream(
+    const lldb::SBStream &stream) const {
+  if (stream.m_opaque_up)
+    return const_cast<lldb::SBStream &>(stream).m_opaque_up.release();
```
This feels a little sketchy to me... `GetOpaqueTypeFromSBStream` is taking ownership of the underlying pointer and returning it to the caller. The `SBStream` that gets passed in is no longer valid. Is this intentional? That would mean this function -- and only this function -- has different semantics from the other `GetOpaqueTypeFromSBClass` functions.

medismailben replied:
`lldb::SBStream` have different semantics than other classes, as it uses a unique_ptr to point to its opaque type and  also `lldb_private::Stream` is an abstract type so the only way to get the underlying object is to do this. Let me know if you can think of a better way of doing it.

bulbazord replied:
I'm having a hard time wrapping my head around the SBStream portion of your patch (everything else makes sense). So we pass in a `const SBStream &` to `GetOpaqueTypeFromSBStream` which that takes ownership of the underlying `Stream *`  and returns it. What happens to the original SBStream argument? Can we guarantee that nothing ever tries to use it again afterwards?
It looks like the intent in `ReverseTransform` is to take ownership of the original user-provided stream and write its contents to some other stream `original_arg` after performing some kind of copy on it. Instead of taking ownership of the underlying stream (and invalidating its argument), why not just use `.get()` instead of `.release()`? The `SBStream` passed in here won't be invalidated and `original_arg` in `ReverseTransform` can still get the data out of the Python `SBStream` object.

What am I missing?

medismailben replied:
Makes sense. Updated the PR.

bulbazord reviewed (approved):
I think all of my comments have been addressed. Thanks!

JDevlieghere commented on lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedPythonInterface.cpp:
```diff
@@ -54,6 +63,31 @@ Status ScriptedPythonInterface::ExtractValueFromPythonObject<Status>(
+  if (lldb::SBEvent *sb_event = reinterpret_cast<lldb::SBEvent *>(
+          python::LLDBSWIGPython_CastPyObjectToSBEvent(p.get())))
+    return m_interpreter.GetOpaqueTypeFromSBEvent(*sb_event);
+  else
+    error.SetErrorString("Couldn't cast lldb::SBEvent to lldb_private::Event.");
```
No else after return. Same below.

JDevlieghere commented on lldb/source/Target/ThreadPlanPython.cpp:
```diff
@@ -77,14 +95,12 @@ bool ThreadPlanPython::ShouldStop(Event *event_ptr) {
-    }
+    auto should_stop_or_err = m_interface->ShouldStop(event_ptr);
+    if (!should_stop_or_err) {
+      llvm::consumeError(should_stop_or_err.takeError());
```
Can we log these errors instead of dropping them?

JDevlieghere reviewed (approved):
LGTM with a few nits. This looks really great with the new interfaces.

medismailben committed:
[lldb] Make use of Scripted{Python,}Interface for ScriptedThreadPlan

This patch makes ScriptedThreadPlan conforming to the ScriptedInterface
& ScriptedPythonInterface facilities by introducing 2
ScriptedThreadPlanInterface & ScriptedThreadPlanPythonInterface classes.

This allows us to get rid of every ScriptedThreadPlan-specific SWIG
method and re-use the same affordances as other scripting offordances,
like Scripted{Process,Thread,Platform} & OperatingSystem.

To do so, this adds new transformer methods for `ThreadPlan`, `Stream` &
`Event`, to allow the bijection between C++ objects and their python
counterparts.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

medismailben commented:
This has been reverted in 6eafe2cb7a3286c1b13eea7d8370374553fe81a9 since [it introduced some test failures](https://lab.llvm.org/buildbot/#/builders/68/builds/62556):

********************
Failed Tests (4):
  lldb-api :: functionalities/completion/TestCompletion.py
  lldb-api :: functionalities/step_scripted/TestStepScripted.py
  lldb-api :: functionalities/thread_plan/TestThreadPlanCommands.py
  lldb-shell :: Commands/command-stop-hook-output.test

DavidSpickett commented:
So in case it was confusing, the revert did fix Arm 32 bit linux as well, but it stayed red for other reasons.

Also I looked into it on the same machine and got:
```
********************
Unresolved Tests (1):
  lldb-api :: functionalities/thread_plan/TestThreadPlanCommands.py

********************
Failed Tests (3):
  lldb-api :: functionalities/completion/TestCompletion.py
  lldb-api :: functionalities/step_scripted/TestStepScripted.py
  lldb-shell :: Commands/command-stop-hook-output.test
```
Same tests as on x86. One failed with:
```
FAIL: LLDB (/home/david.spickett/build-arm/bin/clang-arm) :: test_thread_plan_actions (TestThreadPlanCommands.TestThreadPlanCommands)
======================================================================
ERROR: test_thread_plan_actions (TestThreadPlanCommands.TestThreadPlanCommands)
----------------------------------------------------------------------
TypeError: WrapStepOver.__init__() missing 1 required positional argument: 'dict'
```
And I see that the previous special case Python object maker thingy (can you tell I know nothing about this :) ) did pass on, or make a dict of some sort.

The others seemed to be trying to create a plan, which failed (same reason I expect), then tried to use it which of course failed.

medismailben commented:
@DavidSpickett Checking! Thanks for the heads-up.
